lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -649,6 +675,92 @@ void SPIRVToOCLBase::visitCallSPIRVBuiltin(CallInst *CI, Op OC) {
 }
 
+void SPIRVToOCLBase::visitCallSPIRVOCLExt(CallInst *CI, OCLExtOpKind Kind) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        return OCLExtOpMap::map(Kind);
+      },
+      &Attrs);
+}
+
+void SPIRVToOCLBase::visitCallSPIRVVLoadn(CallInst *CI, OCLExtOpKind Kind) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        std::string Name = OCLExtOpMap::map(Kind);
+        if (ConstantInt *C = dyn_cast<ConstantInt>(Args.back())) {
+          uint64_t NumComponents = C->getZExtValue();
+          std::stringstream SS;
+          SS << NumComponents;
+          Name.replace(Name.find("n"), 1, SS.str());
+        }
+        Args.pop_back();
+        return Name;
+      },
+      &Attrs);
+}
+
+void SPIRVToOCLBase::visitCallSPIRVVStore(CallInst *CI, OCLExtOpKind Kind) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        std::string Name = OCLExtOpMap::map(Kind);
+        if (Kind == OpenCLLIB::Vstore_half_r ||
+            Kind == OpenCLLIB::Vstore_halfn_r ||
+            Kind == OpenCLLIB::Vstorea_halfn_r) {
+          if (ConstantInt *C = dyn_cast<ConstantInt>(Args.back())) {
+            auto RoundingMode =
+                static_cast<SPIRVFPRoundingModeKind>(C->getZExtValue());
+            Name.replace(Name.find("_r"), 2,
+                         std::string("_") +
+                             SPIRSPIRVFPRoundingModeMap::rmap(RoundingMode));
+            Args.pop_back();
+          } else {
+            assert(0 && "Vstore*_r must have FP Rounding Mode specified!");
+          }
+        }
+
+        if (Kind == OpenCLLIB::Vstore_halfn ||
+            Kind == OpenCLLIB::Vstore_halfn_r ||
+            Kind == OpenCLLIB::Vstorea_halfn ||
+            Kind == OpenCLLIB::Vstorea_halfn_r || Kind == OpenCLLIB::Vstoren) {
+          if (auto DataType = dyn_cast<VectorType>(Args[0]->getType())) {
+            auto NumElements = DataType->getElementCount().getValue();
+            assert((NumElements == 2 || NumElements == 3 || NumElements == 4 ||
+                    NumElements == 8 || NumElements == 16) &&
+                   "Unsupported vector size for vstore instruction!");
+            std::stringstream SS;
+            SS << NumElements;
+            Name.replace(Name.find("n"), 1, SS.str());
+          }
+        }
+
+        return Name;
+      },
+      &Attrs);
+}
+
+void SPIRVToOCLBase::visitCallSPIRVPrintf(CallInst *CI, OCLExtOpKind Kind) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  CallInst *NewCI = mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        return OCLExtOpMap::map(OpenCLLIB::Printf);
+      },
+      &Attrs);
```

Comment:
Yes, we need this. `mutateCallInstOCL` creates a new function with variable number of arguments (based on value set in [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1008/files#diff-c9784dd5ceef34a6b4ffee9259d63464942061e70d710f1fdca0eecd91421976R991)).

Reply:
Should we create a variadic function in SPIR-V IR?

Reply:
Our current implementation of printf in SPV-IR doesn't represent it as a variadic function. Only OCL representation is variadic. I'm also wondering whether creating seperate function declaration for all printf calls in SPV-IR is a good solution..

Reply:
> I'm also wondering whether creating seperate function declaration for all printf calls in SPV-IR is a good solution..

It is probably not. I think we might switch to variadic function in the future, but I think it can be done in a separate patch. I'm also not sure if SPIRVWriter will be fine with accepting variadic function in SPIR-V IR (I just have never checked).